src/main/scala/firrtl/backends/experimental/smt/FirrtlToTransitionSystem.scala

```diff
@@ -247,233 +250,116 @@ private class ModuleToTransitionSystem extends LazyLogging {
+        assert(
+          values.length == m.depth,
+          s"Memory ${m.name} of depth ${m.depth} cannot be initialized with an array of length ${values.length}!"
+        )
+        // in order to get a more compact encoding try to find the most common values
+        val histogram = mutable.LinkedHashMap[BigInt, Int]()
+        values.foreach(v => histogram(v) = 1 + histogram.getOrElse(v, 0))
+        val baseValue = histogram.maxBy(_._2)._1
+        val base = ArrayConstant(BVLiteral(baseValue, dataWidth), indexWidth)
+        values.zipWithIndex
+          .filterNot(_._1 == baseValue)
+          .foldLeft[ArrayExpr](base) {
+            case (array, (value, index)) =>
+              ArrayStore(array, BVLiteral(index, indexWidth), BVLiteral(value, dataWidth))
+          }
+      case other => throw new RuntimeException(s"Unsupported memory init option: $other")
     }
 
-  }
-  private class ReadPort(memory: MemInfo, name: String, inputs: String => BVExpr)
-      extends MemPort(memory, name, inputs) {}
-
+  // TODO: add to BV expression library
   private def and(a: BVExpr, b: BVExpr): BVExpr = (a, b) match {
     case (True, True) => True
     case (True, x)    => x
     case (x, True)    => x
     case _            => BVOp(Op.And, a, b)
   }
-  private def or(a: BVExpr, b: BVExpr): BVExpr = BVOp(Op.Or, a, b)
+
   private val True = BVLiteral(1, 1)
-  private val False = BVLiteral(0, 1)
-  private def all(b: Iterable[BVExpr]): BVExpr = if (b.isEmpty) False else b.reduce((a, b) => and(a, b))
-  private def any(b: Iterable[BVExpr]): BVExpr = if (b.isEmpty) True else b.reduce((a, b) => or(a, b))
+  private def checkMem(m: ir.DefMemory): Unit = {
+    assert(m.readLatency == 0, "Expected read latency to be 0. Did you run VerilogMemDelays?")
+    assert(m.writeLatency == 1, "Expected read latency to be 1. Did you run VerilogMemDelays?")
+    assert(
+      m.dataType.isInstanceOf[ir.GroundType],
+      s"Memory $m is of type ${m.dataType} which is not a ground type!"
+    )
+    assert(m.readwriters.isEmpty, "Combined read/write ports are not supported! Please split them up.")
+  }
+
+  private val InfoSeparator = ", "
+  private val InfoPrefix = "@ "
+  private def serializeInfo(info: ir.Info): Option[String] = info match {
+    case ir.NoInfo => None
+    case f: ir.FileInfo => Some(f.escaped)
+    case m: ir.MultiInfo =>
+      val infos = m.flatten
+      if (infos.isEmpty) { None }
+      else { Some(infos.map(_.escaped).mkString(InfoSeparator)) }
+  }
+
+  private[firrtl] val randoms = mutable.LinkedHashMap[String, BVSymbol]()
+  private def makeRandom(baseName: String, width: Int): BVExpr = {
+    // TODO: actually ensure that there cannot be any name clashes with other identifiers
+    val suffixes = Iterator(baseName) ++ (0 until 200).map(ii => baseName + "_" + ii)
+    val name = suffixes.map(s => "RANDOM." + s).find(!randoms.contains(_)).get
+    val sym = BVSymbol(name, width)
+    randoms(name) = sym
+    sym
```

Comment:
This is old code that is going to be removed by a followup PR. Since we are moving all randomness modelling out of the SMT backend and into passes that add `DefRandom` nodes, the backend will no longer need to generate names.